service/src/main/kotlin/fi/espoo/evaka/reports/OccupancyReport.kt

```diff
@@ -273,3 +287,223 @@ private fun Database.Read.calculateOccupancyGroupReport(
         .filter { (isOperationDay, _) -> isOperationDay }
         .map { (_, reportRow) -> reportRow }
 }
+
+private data class Caretakers(
+    @Nested
+    val unit: UnitKey,
+    val date: LocalDate,
+    val caretakers: BigDecimal
+)
+
+private data class Placement(
+    val id: UUID,
+    val childId: UUID,
+    val unitId: UUID,
+    val type: PlacementType,
+    val familyUnitPlacement: Boolean,
+    val period: FiniteDateRange
+)
+
+private data class Child(
+    val id: UUID,
+    val dateOfBirth: LocalDate
+)
+
+private data class ServiceNeed(
+    val placementId: UUID,
+    val occupancyCoefficient: BigDecimal,
+    val period: FiniteDateRange
+)
+
+private data class AssistanceNeed(
+    val childId: UUID,
+    val capacityFactor: BigDecimal,
+    val period: FiniteDateRange
+)
+
+private data class Absence(
+    val childId: UUID,
+    val date: LocalDate
+)
+
+private fun Database.Read.calculateOccupancyUnitReportV2(
+    today: LocalDate,
+    areaId: UUID,
+    queryPeriod: FiniteDateRange,
+    type: OccupancyType
+): List<OccupancyUnitReportResultRow> {
+    if (queryPeriod.start.plusDays(50) < queryPeriod.end) {
+        throw BadRequest("Date range ${queryPeriod.start} - ${queryPeriod.end} is too long. Maximum range is 50 days.")
+    }
+
+    val period =
```

Comment:
Just something to check during testing: if the report is really slow, maybe we want to assert max period length here.

Reply:
There is an assertion right above? 😄

But, it does make more sense to do the assertion for the actual period used in calculations.